improve DataFrame constructors and conversions for Vector and Matrix

bkamins opened the pull request:
Following discussion in https://discourse.julialang.org/t/how-do-dataframes-jl-compare-to-rs-and-interoperability-between-r-and-julia/7387/14 I propose to:
1. Add a constructor of `DataFrame` from a `Matrix` that can take column names argument;
2. Add a method to `convert` allowing conversion from `Vector` to `DataFrame` (so that both the constructor and conversion work and work the same way);

nalimilan commented on src/dataframe/dataframe.jl:
```diff
@@ -802,14 +809,8 @@ function Base.append!(df1::DataFrame, df2::AbstractDataFrame)
-    return DataFrame(cols, Index(gennames(n)))
-end
+Base.convert(::Type{DataFrame}, A::AbstractMatrix) = DataFrame(A)
+Base.convert(::Type{DataFrame}, V::AbstractVector) = DataFrame(V)
```
Note that this method is very different from the previous one as it expects the entries to be columns. I'm not sure we should support it, as it could be confusing compared with a 1-column matrix.

bkamins replied:
To say the truth I do not like `DataFrame(::Vector)` constructor, especially something like `DataFrame([1,2,3])` is very non-intuivite. I pass a column vector and get a row of data.

But if we allow it then it is consistent to provide `convert` also.

Maybe a solution is to restrict the implementation of `DataFrame(columns::AbstractVector` to check internally:
```
all(col -> isa(col, AbstractVector), columns) || throw(ArgumentError("columns must be vectors"))
```

nalimilan replied:
Yeah, it's a bit weird. Maybe we could also require specifying the column names when passing a vector, given that this is a rather low-level constructor.

bkamins replied:
agreed. this means that `convert` should be removed. I add deprecation to calling *vector of vectors* `DataFrame` without passing names.

nalimilan commented on src/dataframe/dataframe.jl:
```diff
@@ -125,6 +126,12 @@ function DataFrame(columns::AbstractVector,
+    DataFrame(Any[columns[:, i] for i in 1:size(columns, 2)], cnames)
+end
+
+
```
One blank line is enough.

bkamins replied:
fixed

nalimilan commented on src/dataframe/dataframe.jl:
```diff
@@ -125,6 +126,12 @@ function DataFrame(columns::AbstractVector,
 end
 
+# if #1308 is merged this sould be refactor to include makeunique
+function DataFrame(columns::AbstractMatrix, cnames::AbstractVector{Symbol} = gennames(size(columns, 2)))
```
Better use the short function syntax for one-liners like this.

bkamins replied:
fixed

bkamins commented:
In R a matrix is actually our `NamedArray` - `data.frame` uses names of columns.
Actually this is something that could be considered do be added in `NamedArrays`.

bkamins commented:
I decided to recommend allowing `DataFrames` constructor (and thus `convert`) from `AbstractVector` only if it contains `AbstractVector`. This is breaking so I put deprecation for the time being.

Other issue is that tests on 0.7 fail (for different reasons - I am not sure if this is something to worry about, but in one case Julia crashed trying to precompile `CategoricalArrays`)

nalimilan commented on src/dataframe/dataframe.jl:
```diff
@@ -122,9 +123,17 @@ end
 function DataFrame(columns::AbstractVector,
                    cnames::AbstractVector{Symbol} = gennames(length(columns)))
+    if !all(col -> isa(col, AbstractVector), columns)
+        # change to throw(ArgumentError("columns argument mutst be a vector of vectors"))
```
"must"

bkamins replied:
fixed

nalimilan commented on src/dataframe/dataframe.jl:
```diff
@@ -122,9 +123,17 @@ end
     return DataFrame(convert(Vector{Any}, columns), Index(convert(Vector{Symbol}, cnames)))
 end
 
+# if #1308 is merged this sould be refactor to include makeunique
```
Let's merge #1308 first.

bkamins replied:
OK. I am pushing the as-is version to have it on GitHub, but I will rebase it when 1308 is merged.

nalimilan commented on src/dataframe/dataframe.jl:
```diff
@@ -120,11 +121,23 @@ function DataFrame(; kwargs...)
 
-function DataFrame(columns::AbstractVector,
-                   cnames::AbstractVector{Symbol} = gennames(length(columns)))
+function DataFrame(columns::AbstractVector)
```
That function should be moved to deprecated.jl.

bkamins replied:
moved

bkamins commented:
rebased, added `makeunique` to `Matrix` constructor and in `Vector` constructor changed that it is required that at least one element is a vector (not all as was earlier).

nalimilan commented:
I'd rather require all entries to be `AbstractVector` as you did originally, not just one of them. That's simpler, and since that's a low-level constructor is not a big deal to require the caller to create a vector.

bkamins commented:
OK - fixed
